controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1199 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"net/http"
+	"os"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/gopkg.in/inconshreveable/log15.v2"
+	"github.com/flynn/flynn/controller/client"
+	ct "github.com/flynn/flynn/controller/types"
+	"github.com/flynn/flynn/controller/utils"
+	"github.com/flynn/flynn/discoverd/client"
+	"github.com/flynn/flynn/host/types"
+	"github.com/flynn/flynn/pkg/attempt"
+	"github.com/flynn/flynn/pkg/cluster"
+	"github.com/flynn/flynn/pkg/httphelper"
+	"github.com/flynn/flynn/pkg/random"
+	"github.com/flynn/flynn/pkg/shutdown"
+	"github.com/flynn/flynn/pkg/status"
+	"github.com/flynn/flynn/pkg/stream"
+)
+
+const (
+	eventBufferSize    int           = 1000
+	serviceName                      = "controller-scheduler"
+	maxJobAttempts     uint          = 30
+	jobAttemptInterval time.Duration = 500 * time.Millisecond
+)
+
+var logger = log15.New("component", "scheduler")
+
+func fnLogger(ctx ...interface{}) log15.Logger {
+	pc := make([]uintptr, 10) // at least 1 entry needed
+	runtime.Callers(2, pc)
+	f := runtime.FuncForPC(pc[0])
+	name := f.Name()
+	name = name[strings.LastIndex(name, ".")+1:]
+	params := []interface{}{"fn", name}
+	params = append(params, ctx...)
+	return logger.New(params...)
+}
+
+type Scheduler struct {
+	utils.ControllerClient
+	utils.ClusterClient
+
+	isLeader         bool
+	changeLeader     chan bool
+	backoffPeriod    time.Duration
+	maxBackoffPeriod time.Duration
+
+	formations  Formations
+	hostStreams map[string]stream.Stream
+	jobs        Jobs
+
+	jobEvents chan *host.Event
+
+	listeners map[chan Event]struct{}
+	listenMtx sync.RWMutex
+
+	ready    bool
+	stop     chan struct{}
+	stopOnce sync.Once
+
+	rectify         chan struct{}
+	syncJobs        chan struct{}
+	syncFormations  chan struct{}
+	hostEvents      chan *discoverd.Event
+	formationEvents chan *ct.ExpandedFormation
+	jobRequests     chan *JobRequest
+	putJobs         chan *ct.Job
+}
+
+func NewScheduler(cluster utils.ClusterClient, cc utils.ControllerClient) *Scheduler {
+	s := &Scheduler{
+		ControllerClient: cc,
+		ClusterClient:    cluster,
+		backoffPeriod:    getBackoffPeriod(),
+		hostStreams:      make(map[string]stream.Stream),
+		jobs:             make(map[string]*Job),
+		formations:       make(Formations),
+		listeners:        make(map[chan Event]struct{}),
+		changeLeader:     make(chan bool),
+		jobEvents:        make(chan *host.Event, eventBufferSize),
+		stop:             make(chan struct{}),
+		rectify:          make(chan struct{}, 1),
+		syncJobs:         make(chan struct{}, 1),
+		syncFormations:   make(chan struct{}, 1),
+		formationEvents:  make(chan *ct.ExpandedFormation, eventBufferSize),
+		hostEvents:       make(chan *discoverd.Event, eventBufferSize),
+		jobRequests:      make(chan *JobRequest, eventBufferSize),
+		putJobs:          make(chan *ct.Job, eventBufferSize),
+	}
+	s.maxBackoffPeriod = 10 * s.backoffPeriod
```

Comment:
The backoff periods increase in powers of 2, so it probably makes sense to cap at a particular power (e.g 32).